README
NAME
    Convert::Scalar - convert between different representations of perl
    scalars

SYNOPSIS
     use Convert::Scalar;

DESCRIPTION
    This module exports various internal perl methods that change the
    internal representation or state of a perl scalar. All of these work
    in-place, that is, they modify their scalar argument. No functions are
    exported by default.

    The following export tags exist:

     :utf8   all functions with utf8 in their name
     :taint  all functions with taint in their name
     :refcnt all functions with refcnt in their name
     :ok     all *ok-functions.

    utf8 scalar[, mode]
        Returns true when the given scalar is marked as utf8, false
        otherwise. If the optional mode argument is given, also forces the
        interpretation of the string to utf8 (mode true) or plain bytes
        (mode false). The actual (byte-) content is not changed. The return
        value always reflects the state before any modification is done.

        This function is useful when you "import" utf8-data into perl, or
        when some external function (e.g. storing/retrieving from a
        database) removes the utf8-flag.

    utf8_on scalar
        Similar to "utf8 scalar, 1", but additionally returns the scalar
        (the argument is still modified in-place).

    utf8_off scalar
        Similar to "utf8 scalar, 0", but additionally returns the scalar
        (the argument is still modified in-place).

    utf8_valid scalar [Perl 5.7]
        Returns true if the bytes inside the scalar form a valid utf8
        string, false otherwise (the check is independent of the actual
        encoding perl thinks the string is in).

    utf8_upgrade scalar
        Convert the string content of the scalar in-place to its
        UTF8-encoded form (and also returns it).

    utf8_downgrade scalar[, fail_ok=0]
        Attempt to convert the string content of the scalar from
        UTF8-encoded to ISO-8859-1. This may not be possible if the string
        contains characters that cannot be represented in a single byte; if
        this is the case, it leaves the scalar unchanged and either returns
        false or, if "fail_ok" is not true (the default), croaks.

    utf8_encode scalar
        Convert the string value of the scalar to UTF8-encoded, but then
        turn off the "SvUTF8" flag so that it looks like bytes to perl
        again. (Might be removed in future versions).

    utf8_length scalar
        Returns the number of characters in the string, counting wide UTF8
        characters as a single character, independent of wether the scalar
        is marked as containing bytes or mulitbyte characters.

    $old = readonly scalar[, $new]
        Returns whether the scalar is currently readonly, and sets or clears
        the readonly status if a new status is given.

    readonly_on scalar
        Sets the readonly flag on the scalar.

    readonly_off scalar
        Clears the readonly flag on the scalar.

    unmagic scalar, type
        Remove the specified magic from the scalar (DANGEROUS!).

    weaken scalar
        Weaken a reference. (See also WeakRef).

    taint scalar
        Taint the scalar.

    tainted scalar
        returns true when the scalar is tainted, false otherwise.

    untaint scalar
        Remove the tainted flag from the specified scalar.

    length = len scalar
        Returns SvLEN (scalar), that is, the actual number of bytes
        allocated to the string value, or "undef", is the scalar has no
        string value.

    scalar = grow scalar, newlen
        Sets the memory area used for the scalar to the given length, if the
        current length is less than the new value. This does not affect the
        contents of the scalar, but is only useful to "pre-allocate" memory
        space if you know the scalar will grow. The return value is the
        modified scalar (the scalar is modified in-place).

    scalar = extend scalar, addlen
        Reserves enough space in the scalar so that addlen bytes can be
        appended without reallocating it. The actual contents of the scalar
        will not be affected. The modified scalar will also be returned.

        This function is meant to make append workloads efficient - if you
        append a short string to a scalar many times (millions of times),
        then perl will have to reallocate and copy the scalar basically
        every time.

        If you instead use "extend $scalar, length $shortstring", then
        Convert::Scalar will use a "size to next power of two, roughly"
        algorithm, so as the scalar grows, perl will have to resize and copy
        it less and less often.

    refcnt scalar[, newrefcnt]
        Returns the current reference count of the given scalar and
        optionally sets it to the given reference count.

    refcnt_inc scalar
        Increments the reference count of the given scalar inplace.

    refcnt_dec scalar
        Decrements the reference count of the given scalar inplace. Use
        "weaken" instead if you understand what this function is fore.
        Better yet: don't use this module in this case.

    refcnt_rv scalar[, newrefcnt]
        Works like "refcnt", but dereferences the given reference first.
        This is useful to find the reference count of arrays or hashes,
        which cnanot be passed directly. Remember that taking a reference of
        some object increases it's reference count, so the reference count
        used by the *_rv-functions tend to be one higher.

    refcnt_inc_rv scalar
        Works like "refcnt_inc", but dereferences the given reference first.

    refcnt_dec_rv scalar
        Works like "refcnt_dec", but dereferences the given reference first.

    ok scalar
    uok scalar
    rok scalar
    pok scalar
    nok scalar
    niok scalar
        Calls SvOK, SvUOK, SvROK, SvPOK, SvNOK or SvNIOK on the given
        scalar, respectively.

  CANDIDATES FOR FUTURE RELEASES
    The following API functions (perlapi) are considered for future
    inclusion in this module If you want them, write me.

     sv_upgrade
     sv_pvn_force
     sv_pvutf8n_force
     the sv2xx family

AUTHOR
     Marc Lehmann <schmorp@schmorp.de>
     http://home.schmorp.de/



Changes
Revision history for Perl extension Convert::Scalar

1.1  Thu Aug 18 20:59:01 CEST 2011
	- added readonly, readonly_on, readonly_off.
        - add extend and len.
        - modernise xs syntax, make some functions more efficient.

1.04  Fri Mar  7 11:32:53 CET 2008
	- work around 5.10 api changes in *ok functions
          (reported by David Aguilar).
 
1.03  Thu Mar  3 18:00:52 CET 2005
	- require perl 5.008.
        - change of contact address.

1.02  Fri Nov 26 22:04:10 CET 2004
	- return copies in all string-modifying functions
          instead of the string itself.

1.01  Sat Apr 24 08:38:21 CEST 2004
	- minor documentation corrections.

1.0   Sat Apr 24 08:21:11 CEST 2004
	- minor documentation corrections.

0.1   Wed Feb 12 21:40:14 CET 2003
	- added *ok functions.

0.09  Mon Sep 23 05:40:41 CEST 2002
	- fixed utf8_length to actually work as documented.

0.08  Fri Jul  5 13:44:17 CEST 2002
	- check for readonly scalars before attempting to do utf8
          conversions. perl will fail badly if not.

0.07  Sun Jan 20 04:50:53 CET 2002
	- added refcnt_inc, refcnt_dec, refcnt etc. methods.

0.06  Thu Jan  4 02:24:23 CET 2001
	- utf8_upgrade now really returns the argument
        - all functions that return the argument now do so correctly
          in all cases, but only when called in a non-void context.

0.05  Thu Dec  7 03:54:50 CET 2000
	- now handles get magic in the utf8 functions. I'm not sure
          wether to SvGETMAGIC the other calls (sv_grow, sv_taint).

0.04  Thu Sep 14 21:00:46 CEST 2000
        - some functions now return the modified scalar, enabling a more
          "functional" style, e.g. return utf2, join "",...;

0.03  Sun Sep  3 04:55:53 CEST 2000
	- added "utf8/utf8_on/utf8_off" functions.
        - added grow as an interface to sv_grow.
        - added "tainted" and fixed taint/untaint to use the published
          interface.

0.02  Sat Jul 15 22:37:36 CEST 2000
	- added missing functions weaken and unmagic to EXPORT_OK.
          This happens when you upload too quickly :*]

0.01  Sat Jul 15 20:15:47 CEST 2000
	- original version; leftover "garbage" from the PApp module



Scalar.pm
=head1 NAME

Convert::Scalar - convert between different representations of perl scalars

=head1 SYNOPSIS

 use Convert::Scalar;

=head1 DESCRIPTION

This module exports various internal perl methods that change the internal
representation or state of a perl scalar. All of these work in-place, that
is, they modify their scalar argument. No functions are exported by default.

The following export tags exist:

 :utf8   all functions with utf8 in their name
 :taint  all functions with taint in their name
 :refcnt all functions with refcnt in their name
 :ok     all *ok-functions.

=over 4

=cut

package Convert::Scalar;

BEGIN {
   $VERSION = '1.1';
   @ISA = qw(Exporter);
   @EXPORT_OK = qw(readonly readonly_on readonly_off weaken unmagic len grow extend);
   %EXPORT_TAGS = (
      taint  => [qw(taint untaint tainted)],
      utf8   => [qw(utf8 utf8_on utf8_off utf8_valid utf8_upgrade utf8_downgrade utf8_encode utf8_decode utf8_length)],
      refcnt => [qw(refcnt refcnt_inc refcnt_dec refcnt_rv refcnt_inc_rv refcnt_dec_rv)],
      ok     => [qw(ok uok rok pok nok niok)],
   );

   require Exporter;
   Exporter::export_ok_tags(keys %EXPORT_TAGS);

   require XSLoader;
   XSLoader::load Convert::Scalar, $VERSION;
}

=item utf8 scalar[, mode]

Returns true when the given scalar is marked as utf8, false otherwise. If
the optional mode argument is given, also forces the interpretation of the
string to utf8 (mode true) or plain bytes (mode false). The actual (byte-)
content is not changed. The return value always reflects the state before
any modification is done.

This function is useful when you "import" utf8-data into perl, or when
some external function (e.g. storing/retrieving from a database) removes
the utf8-flag.

=item utf8_on scalar

Similar to C<utf8 scalar, 1>, but additionally returns the scalar (the
argument is still modified in-place).

=item utf8_off scalar

Similar to C<utf8 scalar, 0>, but additionally returns the scalar (the
argument is still modified in-place).

=item utf8_valid scalar [Perl 5.7]

Returns true if the bytes inside the scalar form a valid utf8 string,
false otherwise (the check is independent of the actual encoding perl
thinks the string is in).

=item utf8_upgrade scalar

Convert the string content of the scalar in-place to its UTF8-encoded form
(and also returns it).

=item utf8_downgrade scalar[, fail_ok=0]

Attempt to convert the string content of the scalar from UTF8-encoded to
ISO-8859-1. This may not be possible if the string contains characters
that cannot be represented in a single byte; if this is the case, it
leaves the scalar unchanged and either returns false or, if C<fail_ok> is
not true (the default), croaks.

=item utf8_encode scalar

Convert the string value of the scalar to UTF8-encoded, but then turn off
the C<SvUTF8> flag so that it looks like bytes to perl again. (Might be
removed in future versions).

=item utf8_length scalar

Returns the number of characters in the string, counting wide UTF8
characters as a single character, independent of wether the scalar is
marked as containing bytes or mulitbyte characters.

=item $old = readonly scalar[, $new]

Returns whether the scalar is currently readonly, and sets or clears the
readonly status if a new status is given.

=item readonly_on scalar

Sets the readonly flag on the scalar.

=item readonly_off scalar

Clears the readonly flag on the scalar.

=item unmagic scalar, type

Remove the specified magic from the scalar (DANGEROUS!).

=item weaken scalar

Weaken a reference. (See also L<WeakRef>).

=item taint scalar

Taint the scalar.

=item tainted scalar

returns true when the scalar is tainted, false otherwise.

=item untaint scalar

Remove the tainted flag from the specified scalar.

=item length = len scalar

Returns SvLEN (scalar), that is, the actual number of bytes allocated to
the string value, or C<undef>, is the scalar has no string value.

=item scalar = grow scalar, newlen

Sets the memory area used for the scalar to the given length, if the
current length is less than the new value. This does not affect the
contents of the scalar, but is only useful to "pre-allocate" memory space
if you know the scalar will grow. The return value is the modified scalar
(the scalar is modified in-place).

=item scalar = extend scalar, addlen

Reserves enough space in the scalar so that addlen bytes can be appended
without reallocating it. The actual contents of the scalar will not be
affected. The modified scalar will also be returned.

This function is meant to make append workloads efficient - if you append
a short string to a scalar many times (millions of times), then perl will
have to reallocate and copy the scalar basically every time.

If you instead use C<extend $scalar, length $shortstring>, then
Convert::Scalar will use a "size to next power of two, roughly" algorithm,
so as the scalar grows, perl will have to resize and copy it less and less
often.

=item refcnt scalar[, newrefcnt]

Returns the current reference count of the given scalar and optionally sets it to
the given reference count.

=item refcnt_inc scalar

Increments the reference count of the given scalar inplace.

=item refcnt_dec scalar

Decrements the reference count of the given scalar inplace. Use C<weaken>
instead if you understand what this function is fore. Better yet: don't
use this module in this case.

=item refcnt_rv scalar[, newrefcnt]

Works like C<refcnt>, but dereferences the given reference first. This is
useful to find the reference count of arrays or hashes, which cnanot be
passed directly. Remember that taking a reference of some object increases
it's reference count, so the reference count used by the C<*_rv>-functions
tend to be one higher.

=item refcnt_inc_rv scalar

Works like C<refcnt_inc>, but dereferences the given reference first.

=item refcnt_dec_rv scalar

Works like C<refcnt_dec>, but dereferences the given reference first.

=item ok scalar

=item uok scalar

=item rok scalar

=item pok scalar

=item nok scalar

=item niok scalar

Calls SvOK, SvUOK, SvROK, SvPOK, SvNOK or SvNIOK on the given scalar,
respectively.

=cut

1;

=back

=head2 CANDIDATES FOR FUTURE RELEASES

The following API functions (L<perlapi>) are considered for future
inclusion in this module If you want them, write me.

 sv_upgrade
 sv_pvn_force
 sv_pvutf8n_force
 the sv2xx family

=head1 AUTHOR

 Marc Lehmann <schmorp@schmorp.de>
 http://home.schmorp.de/

=cut

